chain/find.go

```diff
@@ -58,229 +56,159 @@ func (g *GapIndexer) Run(ctx context.Context) error {
-	log.Debug("finding skipped epochs")
-	reportTime := time.Now()
+-- cross product of heights and tasks
+all_heights_and_tasks_in_range as (
+	select h.height, t.task
+	from interesting_heights h
+	cross join interesting_tasks t
+)
+,
+
+-- all heights from processing reports which were
+-- recorded (by gap_fill or consensus) that it is
+-- a null round with no data to index.
+-- then take cross product of these heights w tasks
+null_round_heights_and_tasks_in_range as (
+	select pr.height, t.task
+	from visor_processing_reports pr
+	cross join interesting_tasks t
+	where pr.status_information = ?3
+	and pr.height between ?0 and ?1
+	group by 1, 2
+)
+,
 
-	var skippedReports []visor.ProcessingReport
-	if err := g.DB.AsORM().ModelContext(ctx, &skippedReports).
-		Order("height desc").
-		Where("status = ?", visor.ProcessingStatusSkip).
-		Where("height >= ?", g.minHeight).
-		Where("height <= ?", g.maxHeight).
-		Select(); err != nil {
-		return nil, xerrors.Errorf("query processing report skips: %w", err)
-	}
-	log.Debugw("executed find skipped epoch query", "count", len(skippedReports))
+-- all heights and tasks which need to be filled
+all_incomplete_heights_and_tasks as (
 
-	gapReport := make([]*visor.GapReport, len(skippedReports))
-	for idx, r := range skippedReports {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		default:
-		}
-		gapReport[idx] = &visor.GapReport{
-			Height:     r.Height,
-			Task:       r.Task,
-			Status:     "GAP",
-			Reporter:   g.name,
-			ReportedAt: reportTime,
-		}
-	}
-	return gapReport, nil
-}
+	-- starting from the set of all heights and tasks
+	-- in our range
+	select height, task
+	from all_heights_and_tasks_in_range
 
-func (g *GapIndexer) findEpochGapsAndNullRounds(ctx context.Context, node GapIndexerLens) (visor.GapReportList, []abi.ChainEpoch, error) {
-	log.Debug("finding epoch gaps and null rounds")
-	reportTime := time.Now()
+		except
 
-	var nullRounds []abi.ChainEpoch
-	var missingHeights []uint64
-	res, err := g.DB.AsORM().QueryContext(
-		ctx,
-		&missingHeights,
-		`
-		SELECT s.i AS missing_epoch
-		FROM generate_series(?, ?) s(i)
-		WHERE NOT EXISTS (SELECT 1 FROM visor_processing_reports WHERE height = s.i);
-		`,
-		g.minHeight, g.maxHeight)
-	if err != nil {
-		return nil, nil, err
-	}
-	log.Debugw("executed find epoch gap query", "count", res.RowsReturned())
+	-- remove all heights and tasks which have at least one OK
+	select height, task
+	from visor_processing_reports
+	where status = ?3
+	and height between ?0 and ?1
 
-	gapReport := make([]*visor.GapReport, 0, len(missingHeights))
-	// walk the possible gaps and query lotus to determine if gap was a null round or missed epoch.
-	for _, gap := range missingHeights {
-		select {
-		case <-ctx.Done():
-			return nil, nil, ctx.Err()
-		default:
-		}
-		gh := abi.ChainEpoch(gap)
-		tsgap, err := node.ChainGetTipSetByHeight(ctx, gh, types.EmptyTSK)
```

Comment:
This shouldn't be removed as doing so changes the behavior of the implementation. Gap find should probably continue to be aware of null rounds vs. gaps. A solution is outlined here: https://github.com/filecoin-project/lily/issues/767#issuecomment-972916153. I'd be fine with that landing as a part of this work, or as a follow-on.

Reply:
As a part of this we should also label gaps resulting from undetected null rounds as `NULLROUND-GAP` or similar.